### PR TITLE
Add support for DeepSpeed

### PR DESCRIPTION
**Pull request by regisss**
# What does this PR do?

This PR adds support for DeepSpeed:
- add `deepspeed.py` and modify the `GaudiTrainer` accordingly,
- add test with GPT2-XL,
- add doc

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: docs/source/deepspeed.mdx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+This argument both indicates that DeepSpeed should be used and points to your DeepSpeed configuration.
+
+
+## Launching your script
+
+Finally, you can run your script as follows:
+```bash
+deepspeed --num_nodes 1 --num_gpus 8 --no_local_rank path_to_my_training_script args
+```
```
Comment (philschmid) [inline docs/source/deepspeed.mdx:98]: you could also use `torchrun` or `torch.distributed`. 

Also, should we add it to the `gaudi_spawn.py`?

File: docs/source/deepspeed.mdx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+-->
+
+# DeepSpeed for HPUs
+
+[DeepSpeed](https://www.deepspeed.ai/) enables you to fit and train larger models on HPUs thanks to various optimizations.
+In particular, you can use the two following ZeRO configurations that have been validated to be fully functioning with Gaudi:
+- ZeRO-1, which partitions the optimizer states across processes.
+- ZeRO-2, which additionnally partitions the gradients across processes so that each process retains only the gradients corresponding to its portion of the optimizer states.
+
+These configurations also enable you to train your model in *bf16* precision.
```
Comment (philschmid) [inline docs/source/deepspeed.mdx:24]: thats not new right?

**Review by philschmid — COMMENTED** [review]
Awesome! Can't wait to play around with it! 
BTW, does that mean all `examples/` are working with DS then?

**Review by libinta — COMMENTED** [review]
@regisss  Can you also mark what are the models you tested with deepspeed? and give the cmd also?

File: docs/source/deepspeed.mdx (hunk 1 above)
Comment (regisss) [reply docs/source/deepspeed.mdx]: Great idea, I just added it in a new commit!

File: docs/source/deepspeed.mdx (hunk 2 above)
Comment (regisss) [reply docs/source/deepspeed.mdx]: You're right, the wording is not great. I meant that DeepSpeed is compatible with Habana Mixed Precision. I changed it.

**Comment by regisss** [comment]
> Awesome! Can't wait to play around with it! BTW, does that mean all `examples/` are working with DS then?

@philschmid yes I tried all of them. There is an issue when using `--load_best_model_at_end` so I removed it from the DeepSpeed example in `examples/image-classification/README.md`. I'm going to investigate this but I think we can merge the PR as is if there is no quick fix.

**Comment by regisss** [comment]
> @regisss Can you also mark what are the models you tested with deepspeed? and give the cmd also?

@libinta I tested all the examples, and almost all the models. I can test all of them tomorrow. I added the commands in the READMEs of each example.

By marking the models, you mean adding a column "DeepSpeed" in the table of the README for instance?

File: docs/source/deepspeed.mdx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+Finally, you can run your script as follows:
+```bash
+deepspeed --num_nodes 1 --num_gpus 8 --no_local_rank path_to_my_training_script args
+```
+
+<Tip warning={true}>
+
+You should set `"use_fused_adam": false` in your Gaudi configuration because it is not compatible with DeepSpeed yet.
+
```
[the remark this answers is not in the document]
Comment (regisss) [reply docs/source/deepspeed.mdx:130]: Yes, any example fails with `use_fused_adam: true` in the Gaudi config and I get the following error:
```
File "/usr/local/lib/python3.8/dist-packages/habana_frameworks/torch/hpex/optimizers/FusedAdamW.py", line 58, in wrap_
    result = step_func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/habana_frameworks/torch/hpex/optimizers/FusedAdamW.py", line 95, in step
    exp_avg, exp_avg_sq = state["exp_avg"], state["exp_avg_sq"]
KeyError: 'exp_avg'
```

**Review by libinta — APPROVED** [review]

**Comment by regisss** [comment]
Tests failed because of a Github PAT that expired. They passed on my AWS instance.

**Pull request merged**